From 5ed72912a2fc1d642ceca3d297e8346bb707b032 Mon Sep 17 00:00:00 2001
From: Johannes Schindelin <johannes.schindelin@gmx.de>
Date: Wed, 7 Dec 2022 15:52:03 +0100
Subject: [PATCH 37/N] Revert "Cygwin: drop requirement to build newlib's
 stdio64"

As per the bug report in https://savannah.gnu.org/bugs/index.php?63480,
2902b3a09e (Cygwin: drop requirement to build newlib's stdio64,
2022-08-03) would require all binaries linking to the MSYS2 runtime to
be rebuilt.

Let's revert that commit instead.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 newlib/configure.host                 | 1 +
 newlib/libc/include/stdio.h           | 4 ++++
 winsup/cygwin/include/cygwin/config.h | 2 ++
 winsup/cygwin/syscalls.cc             | 4 ++--
 4 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/newlib/configure.host b/newlib/configure.host
index 4e4c393..7a5456a 100644
--- a/newlib/configure.host
+++ b/newlib/configure.host
@@ -420,6 +420,7 @@ fi
 case "${host}" in
   *-*-cygwin* | *-*-msys*)
 	posix_dir=posix
+	stdio64_dir=stdio64
 	xdr_dir=xdr
 	;;
   *-*-netware*)
diff --git a/newlib/libc/include/stdio.h b/newlib/libc/include/stdio.h
index ea4ddf3..c75e481 100644
--- a/newlib/libc/include/stdio.h
+++ b/newlib/libc/include/stdio.h
@@ -67,10 +67,14 @@ typedef __FILE FILE;
 # define __FILE_defined
 #endif
 
+#ifdef __CYGWIN__
+typedef _fpos64_t fpos_t;
+#else
 typedef _fpos_t fpos_t;
 #ifdef __LARGE64_FILES
 typedef _fpos64_t fpos64_t;
 #endif
+#endif /* !__CYGWIN__ */
 
 #ifndef _OFF_T_DECLARED
 typedef __off_t off_t;
diff --git a/winsup/cygwin/include/cygwin/config.h b/winsup/cygwin/include/cygwin/config.h
index 2a70832..1d515a6 100644
--- a/winsup/cygwin/include/cygwin/config.h
+++ b/winsup/cygwin/include/cygwin/config.h
@@ -54,6 +54,8 @@ extern inline struct _reent *__getreent (void)
    Just leave them alone. */
 #define _READ_WRITE_RETURN_TYPE _ssize_t
 #define _READ_WRITE_BUFSIZE_TYPE size_t
+#define __LARGE64_FILES 1
+#define __USE_INTERNAL_STAT64 1
 #define __LINUX_ERRNO_EXTENSIONS__ 1
 #define _MB_EXTENDED_CHARSETS_ALL 1
 #define __HAVE_LOCALE_INFO__ 1
diff --git a/winsup/cygwin/syscalls.cc b/winsup/cygwin/syscalls.cc
index 5c7f46a..a27e283 100644
--- a/winsup/cygwin/syscalls.cc
+++ b/winsup/cygwin/syscalls.cc
@@ -4143,7 +4143,7 @@ gethostid (void)
 
 #define ETC_SHELLS "/etc/shells"
 static int shell_index;
-static FILE *shell_fp;
+static struct __sFILE64 *shell_fp;
 
 extern "C" char *
 getusershell ()
@@ -4162,7 +4162,7 @@ getusershell ()
   static char buf[PATH_MAX];
   int ch, buf_idx;
 
-  if (!shell_fp && !(shell_fp = fopen (ETC_SHELLS, "rt")))
+  if (!shell_fp && !(shell_fp = fopen64 (ETC_SHELLS, "rt")))
     {
       if (def_shells[shell_index])
 	return strcpy (buf, def_shells[shell_index++]);
